From f8e1383faa5c7942263a0c5c98028bf8653440f4 Mon Sep 17 00:00:00 2001
From: Chad Monroe <chad.monroe@smartrg.com>
Date: Fri, 20 Aug 2021 08:33:11 -0700
Subject: [PATCH] mt76: set sane default rate if rate info is missing

If no rate can be found via beacon (e.g. hidden network) then
set a sane default of either 11Mbps (2G) or 24Mbps (5G). Using these
higher rates ensures we can still connect to APs which have some of
the lower basic rates disabled.

Signed-off-by: Chad Monroe <chad.monroe@smartrg.com>
---
 mac80211.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/mac80211.c b/mac80211.c
index e282c627..c00fd3ea 100644
--- a/mac80211.c
+++ b/mac80211.c
@@ -1357,9 +1357,16 @@ u16 mt76_calculate_default_rate(struct mt76_phy *phy, int rateidx)
 	int offset = 0;
 	struct ieee80211_rate *rate;
 
-	if (phy->chandef.chan->band == NL80211_BAND_5GHZ)
+	if (phy->chandef.chan->band == NL80211_BAND_2GHZ) {
+		if (rateidx < 0)
+			rateidx = 3; /* 11Mbps */
+	} else if (phy->chandef.chan->band == NL80211_BAND_5GHZ) {
 		offset = 4;
 
+		if (rateidx < 0)
+			rateidx = 4; /* 24Mbps */
+	}
+
 	rate = &mt76_rates[offset + rateidx];
 
 	return rate->hw_value;
-- 
2.30.1 (Apple Git-130)

